File: reference-taskA/aptams/rule_engine/engine.py
"""The deterministic scoring pipeline.

Structure is fixed here; numbers come exclusively from :class:`ScoringTables`. The pipeline
is pure and side-effect-free by design — the same inputs and tables always produce the same
output, which is what lets the attribution and counterfactual layers treat it as ground
truth and lets us generate synthetic cohorts before any real data arrives.

Total = Σ(item score × item weight) + 附加分. The bonus is *added on top of* the weighted
total and the result is **not** capped at 100 — verified against the cohort's own recorded
totals, which reach 104.4 (``analysis/verify_scorecard.py``).
"""

from __future__ import annotations

from aptams.rule_engine.models import ItemScore, StudentAssessment, TotalScore
from aptams.rule_engine.tables import ScoringTables


def score_student(assessment: StudentAssessment, tables: ScoringTables) -> TotalScore:
    """Score one student's 体测 record: measurements → item scores → weighted total → band.

    Every number is read from ``tables``; this function only orchestrates lookups and the
    weighting arithmetic defined by the standard. Raises whatever ``tables`` raises when a
    required value is absent — never substitutes a guess.
    """
    item_scores: list[ItemScore] = []
    weighted_total = 0.0
    total_bonus = 0.0

    for m in assessment.measurements:
        score = tables.score_item(
            item=m.item,
            sex=assessment.sex.value,
            grade=assessment.grade,
            raw=m.value,
        )
        weight = tables.item_weight(item=m.item)
        bonus = tables.bonus_for(
            item=m.item,
            sex=assessment.sex.value,
            grade=assessment.grade,
            raw=m.value,
        )
        item_scores.append(
            ItemScore(
                item=m.item,
                raw=m.value,
                score=score,
                bonus=bonus,
                weight=weight,
                band=band_for_item_score(score, tables),
            )
        )
        weighted_total += score * weight
        total_bonus += bonus

    total = weighted_total + total_bonus
    return TotalScore(
        total=total,
        weighted_total=weighted_total,
        bonus=total_bonus,
        band=tables.grade_band_for_total(total=total),
        band_is_derived=tables.total_band_cutoffs_are_derived,
        items=tuple(item_scores),
    )


def band_for_item_score(score: float, tables: ScoringTables) -> str:
    """The 等级 label the source prints against a per-item score.

    Unlike the total-level cutoffs, these groupings are printed explicitly in the source
    table's leftmost column, so this mapping is exact rather than derived.
    """
    for label, levels in tables.raw.get("item_bands", {}).items():
        if score in levels:
            return label
    # Scores below the lowest printed threshold fall outside every printed 等级 group.
    return "不及格"

File: src/lib/aptams/engine.ts
// The deterministic scoring pipeline — port of aptams/rule_engine/engine.py.
// Raw measurements -> per-item scores -> weighted total -> grade band + 附加分.
// Pure and side-effect-free: same inputs always produce same output, which is
// what lets downstream layers treat it as ground truth.

import {
  BAND_LABEL_EN,
  PASS_THRESHOLD,
  TOTAL_BAND_IS_DERIVED,
  bandForItemScore,
  bandForTotal,
  bonusFor,
  itemMeta,
  itemWeight,
  scoreItem,
} from "./tables";
import type {
  ItemScore,
  RawMeasurement,
  Sex,
  Grade,
  TotalScore,
} from "./types";

export interface ScoreInput {
  sex: Sex;
  grade: Grade;
  measurements: RawMeasurement[];
}

/** Decimal places the recorded totals carry. Port of analysis/scorecard.TOTAL_DECIMALS. */
export const TOTAL_DECIMALS = 1;

/**
 * Round a total to the recorded precision, half away from zero.
 *
 * Port of `analysis/scorecard.round_total`. Binary floats make half-way cases
 * ambiguous — 62.95 is stored slightly below 62.95, so a naive
 * `Math.round(x * 10) / 10` can land either side of the tie depending on the
 * order the seven contributions were summed in.
 *
 * An exact total is always an integer multiple of 0.05: item scores are
 * integers and every weight is a multiple of 0.05, and 附加分 is an integer. So
 * we snap to that grid first, which removes accumulation noise entirely and
 * makes this agree with the Python engine bit-for-bit regardless of summation
 * order. `steps * 0.5` is exactly representable in binary (0.5 is a power of
 * two), so the second rounding step carries no float error of its own.
 */
export function roundTotal(value: number): number {
  if (!Number.isFinite(value)) return value;
  const steps = Math.round(value / 0.05); // snap to the 0.05 grid, half up
  const tenths = Math.round(steps * 0.5); // exact; half up to 1 dp
  return tenths / 10;
}

export function scoreStudent(input: ScoreInput): TotalScore {
  const { sex, grade, measurements } = input;
  const items: ItemScore[] = [];
  let weighted = 0;
  let bonusTotal = 0;

  for (const m of measurements) {
    const meta = itemMeta(m.item);
    const score = scoreItem(m.item, sex, grade, m.value);
    const bonus = bonusFor(m.item, sex, grade, m.value);
    const band = bandForItemScore(score);
    const weight = itemWeight(m.item);
    items.push({
      item: m.item,
      raw: m.value,
      unit: meta.unit,
      score,
      bonus,
      band,
      label_zh: meta.label_zh,
      label_en: meta.label_en,
      weight,
    });
    weighted += score * (weight / 100);
    bonusTotal += bonus;
  }

  // 附加分 is added on top of the weighted total and the result is NOT capped
  // at 100 — verified against the cohort's own recorded totals, which reach
  // 104.4 (analysis/verify_scorecard.py). Each item's bonus is already bounded
  // by its own max_bonus in the scoring table; there is no additional
  // cross-item cap in the standard, so none is applied here.
  const total = roundTotal(weighted + bonusTotal);
  const base_total = roundTotal(weighted);
  const band = bandForTotal(total);

  return {
    total,
    base_total,
    bonus_total: bonusTotal,
    band,
    band_is_derived: TOTAL_BAND_IS_DERIVED,
    pass: total >= PASS_THRESHOLD,
    items,
  };
}

export { BAND_LABEL_EN, PASS_THRESHOLD, TOTAL_BAND_IS_DERIVED };
